src/codegen/constant_folding.rs

```diff
@@ -604,37 +658,82 @@ fn expression(
     }
 }
 
-fn bigint_to_expression(loc: &Loc, ty: &Type, n: BigInt) -> (Expression, bool) {
-    let n = match ty {
+fn bigint_to_expression(
+    loc: &Loc,
+    ty: &Type,
+    value: BigInt,
+    overflowing: bool,
+    ns: &mut Namespace,
+) -> (Expression, bool) {
+    let value = match ty {
         Type::Uint(bits) => {
-            if n.bits() > *bits as u64 {
-                let (_, mut bs) = n.to_bytes_le();
+            if value.sign() == Sign::Minus {
+                if !overflowing {
+                    ns.diagnostics.push(Diagnostic::error(
+                        *loc,
+                        format!(
+                            "arithmetic overflow: {value} does not fit into {}",
+                            ty.to_string(ns)
+                        ),
+                    ));
+                }
+                let mut bs = value.to_signed_bytes_le();
+                bs.resize(*bits as usize / 8, 0xff);
+
+                BigInt::from_bytes_le(Sign::Plus, &bs)
+            } else if value.bits() > *bits as u64 || value.sign() == Sign::Minus {
+                if !overflowing {
+                    ns.diagnostics.push(Diagnostic::error(
+                        *loc,
+                        format!(
+                            "arithmetic overflow: {value} does not fit into {}",
+                            ty.to_string(ns)
+                        ),
+                    ));
+                }
+                let (_, mut bs) = value.to_bytes_le();
+                bs.truncate(*bits as usize / 8);
+
+                BigInt::from_bytes_le(Sign::Plus, &bs)
+            } else if value.bits() > *bits as u64 {
```

Comment:
This branch is never reached, because the condition is the same as in the previous.

Reply:
Good point, removed